### PR TITLE
Polyhedron demo: plugins in sub-directories

### PR DESCRIPTION
**Pull request by maxGimeno**
This pull request puts all the plugins into sub-directories, which makes the Polyhedron_demo a little cleaner. 
 Moreover, it adds the possibility to load a plugin manually at run time. 
(ATTENTION : you might have to clean your polyhedron_demo build directory from all the plugins after merging this branch, or some actions in the Operation menu may be in double.)

### CODE REVIEW

File: Polyhedron/demo/Polyhedron/CMakeLists.txt [hunk 1]
```diff
@@ -387,122 +300,41 @@ if(CGAL_Qt5_FOUND AND Qt5_FOUND AND OPENGL_FOUND AND QGLVIEWER_FOUND)
...
-  target_link_libraries(shortest_path_plugin scene_polyhedron_item scene_polylines_item scene_polyhedron_selection_item scene_polyhedron_shortest_path_item scene_basic_objects)
-
-  qt5_wrap_ui( exampleUI_FILES Polyhedron_demo_example_plugin.ui dock_example.ui)
-  qt5_wrap_ui( example_dockUI_FILES dock_example.ui)
-  polyhedron_demo_plugin(example_plugin Polyhedron_demo_example_plugin ${exampleUI_FILES} ${example_dockUI_FILES})
-  target_link_libraries(example_plugin scene_polyhedron_item)
+  add_subdirectory(Selection_plugin)
+  add_subdirectory(Shortest_path_plugin)
+  add_executable(example_plugin IMPORTED)
+  set_property(TARGET example_plugin PROPERTY IMPORTED_LOCATION "${CMAKE_CURRENT_SOURCE_DIR}/../../../Example_plugin/example_plugin.cmake")
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/CMakeLists.txt]: I think that those two lines about `example_plugin` are:
  - not working:
    - the `example_plugin` is not an executable,
    - its `IMPORTED_LOCATION` cannot be a CMake file.
  - useless: the polyhedron demo does not need the list of plugins at compile time, they are discovered at runtime.

Comment (maxGimeno) [inline Polyhedron/demo/Polyhedron/CMakeLists.txt]: They are useless. I thought I removed them.

**Comment by lrineau** [comment]
This branch will quite naturally have a lot of conflicts as soon as there are changed in the plugins. Could it be tested and integrated soon?

Cc: @sloriot, @sgiraudot

**Comment by sgiraudot** [comment]
Agreed. I can test it as soon as needed if the conflicts with cgal/master are solved.

(If needed, I can solve them myself as some probably come from my recent merged PR on the surface reconstruction plugin. Just tell me.)

**Comment by sgiraudot** [comment]
I have the following errors when doing "cmake":

-- The C compiler identification is GNU 5.2.1
-- The CXX compiler identification is GNU 5.2.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build type: Release
-- USING CXXFLAGS = '-Wall -Wextra -frounding-math -O3 -DNDEBUG'
-- USING EXEFLAGS = ' '
-- Targetting Unix Makefiles
-- Using /usr/bin/c++ compiler.
-- Code taken from Git branch: master
-- Requested component: Qt5
-- Requested component: MPFR
-- Requested component: GMP
-- Build type: Release
-- USING CXXFLAGS = '-Wall -Wextra -frounding-math -O3 -DNDEBUG'
-- USING EXEFLAGS = ' '
-- Targetting Unix Makefiles
-- Using /usr/bin/c++ compiler.
-- Code taken from Git branch: master
-- Requested component: Qt5
-- Requested component: MPFR
-- Requested component: GMP
-- Found QGLViewer: /usr/lib/x86_64-linux-gnu/libQGLViewer.so
-- Found Eigen3: /usr/local/include/eigen3 (found suitable version "3.2.90", minimum required is "3.1.0") 
-- Performing Test TBB_without_pthread
-- Performing Test TBB_without_pthread - Failed
-- Performing Test TBB_with_pthread
-- Performing Test TBB_with_pthread - Success
-- Found Intel TBB
-- Boost version: 1.58.0
-- Found the following Boost libraries:
--   thread
--   system
-- Configuring done
CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Function_dialog.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Io_implicit_function_plugin/CMakeLists.txt:3 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Point_inside_polyhedron_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Point_inside_polyhedron_plugin/CMakeLists.txt:3 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Polyhedron_slicer_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Polyhedron_slicer_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Mesh_segmentation_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Mesh_segmentation_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Point_set_selection_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Point_set_selection_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Hole_filling_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Hole_filling_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Fairing_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Fairing_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Selection_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Selection_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error at polyhedron_demo_macros.cmake:25 (add_library):
  Cannot find source file:

    /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build/ui_Shortest_path_widget.h

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
Call Stack (most recent call first):
  Shortest_path_plugin/CMakeLists.txt:4 (polyhedron_demo_plugin)


CMake Error: CMake can not determine linker language for target: io_implicit_function_plugin
CMake Error: Cannot determine link language for target "io_implicit_function_plugin".
CMake Error: CMake can not determine linker language for target: point_inside_polyhedron_plugin
CMake Error: Cannot determine link language for target "point_inside_polyhedron_plugin".
CMake Error: CMake can not determine linker language for target: polyhedron_slicer_plugin
CMake Error: Cannot determine link language for target "polyhedron_slicer_plugin".
CMake Error: CMake can not determine linker language for target: mesh_segmentation_plugin
CMake Error: Cannot determine link language for target "mesh_segmentation_plugin".
CMake Error: CMake can not determine linker language for target: point_set_selection_plugin
CMake Error: Cannot determine link language for target "point_set_selection_plugin".
CMake Error: CMake can not determine linker language for target: hole_filling_plugin
CMake Error: Cannot determine link language for target "hole_filling_plugin".
CMake Error: CMake can not determine linker language for target: fairing_plugin
CMake Error: Cannot determine link language for target "fairing_plugin".
CMake Error: CMake can not determine linker language for target: selection_plugin
CMake Error: Cannot determine link language for target "selection_plugin".
CMake Error: CMake can not determine linker language for target: shortest_path_plugin
CMake Error: Cannot determine link language for target "shortest_path_plugin".
-- Generating done
-- Build files have been written to: /home/simon/CGAL/My_sandbox/Polyhedron/demo/Polyhedron/build

**Comment by sgiraudot** [comment]
Everything seems okay and compiles without any error now.

**Comment by sloriot** [comment]
@maxGimeno please rebase your branch with the latest master.

**Comment by lrineau** [comment]
Sébastien, can you please merge this PR in `integration` for tonight?

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-38

**Comment by lrineau** [comment]
This PR was actually not merged into `integration`. And it does conflict with `integration`.

**Comment by sloriot** [comment]
Merged in CGAL-4.8-Ic-39

**Comment by sloriot** [comment]
There are warnings on windows complaining about the length of object files. One solution would be to remove `Polyhedron_demo_xxxx.cpp` in `Xxxx.cpp`

cc @maxGimeno

**Comment by afabri** [comment]
In fact it would make sense to shorten the names
if each plugin is in its own subdirectory,

andreas

**Comment by sloriot** [comment]
Would it make sense to rename each plugin directory with the name of the packages?
@maxGimeno: Also could we have more that one plugin per directory?
About filenames, I agree with Andreas, `plugin.h`, `plugin.cpp`, `widget.ui`, etc are good to me when there is only one plugin per directory.

**Comment by maxGimeno** [comment]
Don't the subdirectories already have the packages names ? I agree to
rename the files inside them, it soes make sense.

2015-11-05 17:50 GMT+01:00 Sebastien Loriot <notifications@github.com>:

**Comment by afabri** [comment]
Note that the produced dlls must have different names
if they are in one bin directory, so plugin.cpp as name for
the plugins themselves is probably a bad idea.

I only suggested to remove the Polyhedron_demo prefix

andreas

On 06/11/2015 08:14, Maxime GIMENO wrote:
> Don't the subdirectories already have the packages names ? I agree to
> rename the files inside them, it soes make sense.
>
> 2015-11-05 17:50 GMT+01:00 Sebastien Loriot <notifications@github.com>:
>
>  > Would it make sense to rename each plugin directory with the name of the
>  > packages?
>  > @maxGimeno <https://github.com/maxGimeno>: Also could we have more that
>  > one plugin per directory?
>  > About filenames, I agree with Andreas, plugin.h, plugin.cpp, widget.ui,
>  > etc are good to me when there is only one plugin per directory.
>  >
>  > —
>  > Reply to this email directly or view it on GitHub
>  > <https://github.com/CGAL/cgal/pull/430#issuecomment-154119784>.
>  >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CGAL/cgal/pull/430#issuecomment-154331940>.
>

-- 
Andreas Fabri, PhD
Chief Officer, GeometryFactory
Editor, The CGAL Project

phone: +33.492.954.912    skype: andreas.fabri

**Comment by lrineau** [comment]
The source file can be `plugin.cpp`. The name of the plugin will be decided anyway by the content of the `CMakeLists.txt`.

**Comment by maxGimeno** [comment]
Sebastien and I thougt about re-grouping the plugins the same way Jane did
with the submenus, in a subdirectory corresponding to their CGAL include
folder. The subdirectory wouldn't end with _plugin and the plugin files
wouldn't start with Polyhedron_demo. This way it would be coherent with the
Operations menu and the names should be short enough for windows.

2015-11-06 10:24 GMT+01:00 Laurent Rineau <notifications@github.com>:

**Comment by sgiraudot** [comment]
I agree with Maxime.

Also, it is a minor problem, but I think having all files named "plugin.cpp" might be annoying when editing several at the same time in a text editor...

**Comment by sloriot** [comment]
@maxGimeno Could you please refresh the branch so that we can test it tomorrow?

**Comment by sloriot** [comment]
I have some runtime warnings:
```
QObject::connect: No such signal Scene::itemAboutToBeDestroyed(Scene_item*)
QObject::connect:  (receiver name: 'MainWindow')
Enable scripts.
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron"...
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins"...
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron"...
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/AABB_tree"...
### Loading " libcut_plugin.so "...  success
QObject::connect: No such signal Scene::itemAboutToBeDestroyed(Scene_item*)
QObject::connect:  (receiver name: 'shortest_path_plugin')
### Loading " libshortest_path_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Camera_position"...
### Loading " libcamera_positions_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Convex_decomposition"...
### Loading " libnef_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Convex_hull"...
### Loading " libconvex_hull_plugin.so "...  success
### Loading " libkernel_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/IO"...
### Ignoring plugin "libgocad_plugin.so".
    + Loading Function "libp_klein_function_plugin.so"...
    + Loading Function "libp_sphere_function_plugin.so"...
    + Loading Function "libp_tanglecube_function_plugin.so"...
### Loading " libio_implicit_function_plugin.so "...  success
### Loading " libnef_io_plugin.so "...  success
### Loading " liboff_plugin.so "...  success
### Loading " liboff_to_nef_plugin.so "...  success
### Loading " liboff_to_xyz_plugin.so "...  success
### Loading " libply_to_xyz_plugin.so "...  success
### Loading " libpolylines_io_plugin.so "...  success
### Loading " libselection_io_plugin.so "...  success
### Loading " libstl_plugin.so "...  success
### Loading " libxyz_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Mesh_3_plugin"...
### Loading " libdetect_sharp_edges_plugin.so "...  success
### Loading " libio_image_plugin.so "...  success
### Loading " libmesh_3_optimization_plugin.so "...  success
### Loading " libmesh_3_plugin.so "...  success
QObject::connect: No such signal Viewer::pointSelected(QPoint)
QObject::connect:  (sender name:   'viewer')
### Loading " libmesh_3_volume_planes_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Operations_on_polyhedra"...
### Loading " libcorefinement_plugin.so "...  success
### Loading " libintersection_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/PCA"...
### Loading " libpca_plugin.so "...  success
### Loading " libtransform_polyhedron_plugin.so "...  success
### Loading " libtrivial_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/PMP"...
### Loading " libfairing_plugin.so "...  success
QObject::connect: No such signal Scene::itemAboutToBeDestroyed(Scene_item*)
QObject::connect:  (receiver name: 'hole_filling_plugin')
### Loading " libhole_filling_plugin.so "...  success
### Loading " libhole_filling_polyline_plugin.so "...  success
### Loading " libinside_out_plugin.so "...  success
### Loading " libisotropic_remeshing_plugin.so "...  success
### Loading " libjet_fitting_plugin.so "...  success
### Loading " libjoin_and_split_polyhedra_plugin.so "...  success
QObject::connect: No such signal Scene::itemAboutToBeDestroyed(Scene_item*)
QObject::connect:  (receiver name: 'mean_curvature_flow_skeleton_plugin')
### Loading " libmean_curvature_flow_skeleton_plugin.so "...  success
### Loading " liborient_soup_plugin.so "...  success
### Loading " libpoint_inside_polyhedron_plugin.so "...  success
### Loading " libpolyhedron_slicer_plugin.so "...  success
### Loading " libpolyhedron_stitching_plugin.so "...  success
### Loading " librepair_polyhedron_plugin.so "...  success
QObject::connect: No such signal Scene::itemAboutToBeDestroyed(Scene_item*)
QObject::connect:  (receiver name: 'selection_plugin')
### Loading " libselection_plugin.so "...  success
### Loading " libself_intersection_plugin.so "...  success
### Loading " libtriangulate_facets_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Point_set"...
### Loading " libfeatures_detection_plugin.so "...  success
### Loading " libnormal_estimation_plugin.so "...  success
### Loading " libpoint_set_average_spacing_plugin.so "...  success
### Loading " libpoint_set_bilateral_smoothing_plugin.so "...  success
### Loading " libpoint_set_outliers_removal_plugin.so "...  success
### Loading " libpoint_set_selection_plugin.so "...  success
### Loading " libpoint_set_shape_detection_plugin.so "...  success
### Loading " libpoint_set_simplification_plugin.so "...  success
### Loading " libpoint_set_smoothing_plugin.so "...  success
### Loading " libpoint_set_upsampling_plugin.so "...  success
### Loading " libpoint_set_wlop_plugin.so "...  success
### Loading " libsurface_reconstruction_plugin.so "...  success
### Loading " libvcm_normal_estimation_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Subdivision_methods"...
### Loading " libsubdivision_methods_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Surface_mesh"...
QObject::connect: No such signal Scene::itemAboutToBeDestroyed(Scene_item*)
QObject::connect:  (receiver name: 'mesh_segmentation_plugin')
### Loading " libmesh_segmentation_plugin.so "...  success
### Loading " libmesh_simplification_plugin.so "...  success
### Loading " libparameterization_plugin.so "...  success
### Loading " libremeshing_plugin.so "...  success
# Looking for plugins in directory "/home/sloriot/CGAL/git/integration/build/test_clang/demo/Polyhedron/Plugins/Surface_modeling"...
### Loading " libedit_polyhedron_plugin.so "...  success
Succesfully loaded shader func 'VisibilityVP' in file 'Raycasting.glsl'
Succesfully loaded shader func 'VisibilityFP' in file 'Raycasting.glsl'
Succesfully loaded shader func 'AttributeVP' in file 'Raycasting.glsl'
Succesfully loaded shader func 'AttributeFP' in file 'Raycasting.glsl'
Succesfully loaded shader func 'Finalization' in file 'Finalization.glsl'
Convex hull...ok (1 ms)
Simplify...ok (1 ms, 6 edges)
Simplify...ok (1 ms, 6 edges)
Write setting... done.
```

**Comment by sloriot** [comment]
There are some dll import/export issues on windows.

**Comment by sloriot** [comment]
All the issues reported was fixed and tested locally on windows and Linux (gcc and clang).

**Comment by lrineau** [comment]
The PR is also merged. Please update https://cgal.geometryfactory.com/CGAL/Members/wiki/Staging_Area/Polyhedron_demo_enhancement

**Pull request merged**